examples/golang/waku.go

```diff
@@ -0,0 +1,108 @@
+package main
+
+/*
+	#cgo LDFLAGS: -L./ -lwaku -Wl,--allow-multiple-definition
+
+	#include "libwaku.h"
+
+	static NimStringDesc wakuString;
```

Comment:
the nim string layout is not stable between Nim versions and it is part of the garbage collected set of types - throughout, we should use `cstring` for FFI instead

Reply:
Morning and thanks for the comment @arnetheduck !

The thing is that, for example, the next Nim C-exposed proc:
```code
proc waku_new(config: ConfigNode,
              jsonResp: var string): bool
              {.dynlib, exportc.} =
...
```

Is automatically translated to the next when creating the `libwaku.so` library (`nim c --app:lib ...`)
```code
N_LIB_IMPORT N_CDECL(NIM_BOOL, waku_new)(ConfigNode* config, NimStringDesc** jsonResp);
```

And the `jsonResp` variable should be a `var string`. This approach is followed by all the functions that need to return a variable amount of data out of the library. And this is achieved by allocating the "string" variable outside the library and then offering it to the library to get populated from within it.

What would be the best approach to do the same but using `cstring` instead?
Also, I want to avoid allocating memory inside the library.

Reply:
You would use the same approach but use `jsonResp: cstring` (or, more likely, `jsonResp: ptr cchar`) - there's also the question of how to communicate the length: this would again be a separate argument, ie `jsonResp: ptr cchar, jsonRespLen: csize_t`).

> Is automatically translated to the next when creating the libwaku.so library (nim c --app:lib ...)

this may sound convenient initially, but it is a problem for the aforementioned reason: between Nim versions this declaration will not be stable. For example, depending on what `ConfigNode` is and its size, it may be trnslated to a pointer or a value and you don't control which.

For C FFI, `ConfigNode` should be an object (not a `ref object`) and the exported API declaration would be `config: ptr ConfigNode` to indicate that it must be passed by pointer.